File: 7.lifecycle/src/lifecycle_py_pkg/lifecycle_py_pkg/sensor_monitor.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from std_msgs.msg import String


class SensorMonitor(Node):
    def __init__(self):
        super().__init__("sensor_monitor")
        self.declare_parameter("topic_name", "sensor_data")
        topic_name = (
            self.get_parameter("topic_name").get_parameter_value().string_value
        )

        self.received_count = 0
        self.sub = self.create_subscription(
            String, topic_name, self.topic_callback, 10
        )
        self.get_logger().info(
            f"SensorMonitor started. Listening to '{topic_name}' for lifecycle publisher stream..."
        )

    def topic_callback(self, msg: String):
        self.received_count += 1
        self.get_logger().info(
            f"[MONITOR #{self.received_count}] Live message received: '{msg.data}'"
        )


def main(args=None):
    rclpy.init(args=args)
    node = SensorMonitor()
    try:
        rclpy.spin(node)
    except (KeyboardInterrupt, rclpy.executors.ExternalShutdownException):
        pass
    finally:
        node.destroy_node()
        rclpy.try_shutdown()


if __name__ == "__main__":
    main()

File: 7.lifecycle/src/lifecycle_cpp_pkg/src/sensor_monitor.cpp
#include <memory>
#include <string>

#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/string.hpp"

class SensorMonitor : public rclcpp::Node
{
public:
  SensorMonitor() : Node("sensor_monitor"), received_count_(0)
  {
    this->declare_parameter<std::string>("topic_name", "sensor_data");
    std::string topic_name = this->get_parameter("topic_name").as_string();

    sub_ = this->create_subscription<std_msgs::msg::String>(
      topic_name,
      10,
      std::bind(&SensorMonitor::topic_callback, this, std::placeholders::_1));

    RCLCPP_INFO(
      get_logger(),
      "SensorMonitor started. Listening to '%s' for lifecycle publisher stream...",
      topic_name.c_str());
  }

private:
  void topic_callback(const std_msgs::msg::String::SharedPtr msg)
  {
    received_count_++;
    RCLCPP_INFO(
      get_logger(),
      "[MONITOR #%lu] Live message received: '%s'",
      received_count_, msg->data.c_str());
  }

  uint64_t received_count_;
  rclcpp::Subscription<std_msgs::msg::String>::SharedPtr sub_;
};

int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);
  auto node = std::make_shared<SensorMonitor>();
  rclcpp::spin(node);
  rclcpp::shutdown();
  return 0;
}
